compiler/passes/buildDefaultFunctions.cpp

```diff
@@ -307,6 +307,69 @@ static void fixupAccessor(AggregateType* ct, Symbol *field,
     fn->addFlag(FLAG_REF_TO_CONST_WHEN_CONST_THIS);
 }
 
+// Resetting a union to store a different field requires two steps:
+//   1. deinit anything that was stored
+//   2. default-init the requested field
+// If newField == NULL, skip step 2.
+static BlockStmt* buildResetUnionField(Symbol* _this,
+                                       AggregateType* ct,
+                                       Symbol* newField) {
+
+  INT_ASSERT(ct->isUnion());
+
+  BlockStmt* block = new BlockStmt();
+
+  // First, deinitialize whatever is there.
+  for_fields(field, ct) {
+    Symbol* fieldNameSym = new_CStringSymbol(field->name);
+
+    CallExpr* checkId = new CallExpr("==",
+                              new CallExpr(PRIM_GET_UNION_ID, _this),
+                              new CallExpr(PRIM_FIELD_NAME_TO_NUM,
+                                           ct->symbol,
+                                           fieldNameSym));
+
+    BlockStmt* deinitBlock = new BlockStmt();
+    VarSymbol* tmp = newTemp("union_reset_dnt");
```

Comment:
Maybe insert a comment that the "dnt" stands for "deinit"?  It took me a minute.  I also wouldn't object to just typing "deinit" out fully